### PR TITLE
feat: Add ilike and trigram similarity search to pagination

### PR DESCRIPTION
**Pull request by wschurman**
# Why

This PR adds search capabilities to the existing cursor-based pagination system added in #422 which only supported orderBy ordering.

This adds:
1. Case-insensitive pattern matching (ILIKE) - Useful for basic text search across entity fields
2. Trigram similarity search - Provides fuzzy matching capabilities for more advanced search use cases like handling typos or finding similar text

This enables building user-facing search features while maintaining the benefits of cursor-based pagination (stable results, efficient queries).

# How

The implementation introduces a unified PaginationSpecification that supports three strategies:

  1. Standard pagination - The existing orderBy-based pagination
  2. ILIKE search - Pattern matching with automatic wildcard escaping
  3. Trigram search - PostgreSQL trigram similarity with configurable threshold

  Key implementation details:
  - Search terms are properly parameterized to prevent SQL injection
  - ILIKE special characters (%, _, \) are escaped to prevent pattern injection
  - Results maintain stable cursor-based pagination with proper ordering:
    - ILIKE: Ordered by search fields, then ID
    - Trigram: Ordered by exact match priority, similarity score, optional extra fields, then ID

# Test Plan

  The PR includes comprehensive test coverage:

  - Unit tests for both AuthorizationResultBasedKnexEntityLoader and
  EnforcingKnexEntityLoader
  - Integration tests covering:
    - Forward/backward pagination with search
    - Multi-field search
    - Case-insensitive matching
    - Trigram similarity with various thresholds
    - Cursor stability across pages
    - Edge cases (empty results, special characters)
    - Combined with WHERE clauses
    - Security validation ensuring proper escaping and parameterization

All existing tests pass, confirming backward compatibility for standard pagination when using the new API.

### CODE REVIEW

**Comment by wschurman** [comment]
* **#444** <a href="https://app.graphite.com/github/pr/expo/entity/444?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#446](https://github.com/expo/entity/pull/446) <a href="https://app.graphite.com/github/pr/expo/entity/446?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#448](https://github.com/expo/entity/pull/448) <a href="https://app.graphite.com/github/pr/expo/entity/448?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#443** <a href="https://app.graphite.com/github/pr/expo/entity/443?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#442** <a href="https://app.graphite.com/github/pr/expo/entity/442?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#441** <a href="https://app.graphite.com/github/pr/expo/entity/441?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#438** <a href="https://app.graphite.com/github/pr/expo/entity/438?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#437** <a href="https://app.graphite.com/github/pr/expo/entity/437?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#436** <a href="https://app.graphite.com/github/pr/expo/entity/436?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#431** <a href="https://app.graphite.com/github/pr/expo/entity/431?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/expo/entity/431?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by wschurman** [comment]
(for reviewers that are extra curious, you can see the evolution of this over time by seeing the versions on graphite. claude helped quite a lot but required a massive amount of manual intervention)

File: packages/entity-database-adapter-knex/src/internal/EntityKnexDataManager.ts [hunk 1]
```diff
@@ -370,17 +487,251 @@ export class EntityKnexDataManager<
...
+
+  private buildTrigramExactMatchCaseExpression(
+    search: DataManagerSearchSpecification<TFields>,
+    direction: PaginationDirection,
+    tableAlias?: typeof CURSOR_ROW_TABLE_ALIAS,
+  ): SQLFragment {
+    const ilikeConditions = this.buildILikeConditions(search, tableAlias);
+    return sql`CASE WHEN ${SQLFragment.join(ilikeConditions, ' OR ')} THEN ${raw(
+      direction === PaginationDirection.FORWARD ? '1' : '0',
+    )} ELSE ${raw(direction === PaginationDirection.FORWARD ? '0' : '1')} END`;
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/entity-database-adapter-knex/src/internal/EntityKnexDataManager.ts]: Good catch. Technically it worked because the order was also flipped in `buildTrigramExactMatchPriority` based on direction, but the flip wasn't necessary. Updated and existing tests still pass.

From claude:
```
You're right. Look at buildTrigramExactMatchPriority (line 672): it already flips both the CASE
  values and the sort direction. That's redundant — CASE WHEN ... THEN 1 ELSE 0 END DESC and CASE WHEN
   ... THEN 0 ELSE 1 END ASC produce the same ordering.

  And buildTrigramExactMatchCaseExpression (used in the cursor comparison) also flips the values based
   on direction, but the cursor comparison operator is already flipped at line 549 (< vs >).

  The direction-aware value flipping is doing nothing in both places. The sort direction / comparison
  operator already handles it. You should just always use 1 for match and 0 for no-match, and let
  DESC/ASC and </> do the work.
```

**Comment by wschurman** [comment]
### Merge activity

* **Feb 16, 8:58 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/expo/entity/431).
* **Feb 16, 8:58 PM UTC**: @wschurman merged this pull request with [Graphite](https://app.graphite.com/github/pr/expo/entity/431).
